rbd-wnbd: wait for disks to become available

petrutlucian94 opened the pull request:
After a WNBD mapping is created, the driver informs Storport that the bus changed, expecting it to rescan the bus and expose the disk. This is an asynchronous process and it usually takes a matter of milliseconds, up to a few seconds under significant load.

The fsx librbd test ocasionally fails to open the disk as it isn't ready yet. Unlike the Python rbd-wnbd test, it doesn't perform any polling.

For convenience, WNBD now includes a helper function called WnbdPollDiskNumber. We're going to use it to wait for the new disk attachments to become available.

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

petrutlucian94 commented:
I've triggered a new Windows image build in order to pick up the latest wnbd changes: https://jenkins.ceph.com/view/Windows/job/ceph-windows-image-build/28/.

petrutlucian94 commented:
jenkins test windows

idryomov commented on src/tools/rbd_wnbd/rbd_wnbd.cc:
```diff
@@ -1074,6 +1074,9 @@ Map options:
   --read-only             Map read-only
+  --wait-online           Wait for the disk to become available.
+  --wait-online-timeout   Specifies how long to wait for the disk to become
+                          available (seconds). Default: 15
```
`DEFAULT_IMAGE_MAP_TIMEOUT` is defined to 20.

petrutlucian94 replied:
Whops, thanks.

idryomov commented on src/tools/rbd_wnbd/rbd_wnbd.h:
```diff
@@ -69,6 +70,9 @@ struct Config {
   bool remap_failure_fatal = false;
   bool adapter_monitoring_enabled = false;
 
+  bool wait_disk_online = false;
```
Would anyone want a WNBD mapping that isn't online?  It seems like `--wait-online` behavior should be the default (perhaps even with a nice bonus of being able to output the Windows disk path on stdout like it's done on Linux)?

idryomov commented on src/tools/rbd_wnbd/rbd_wnbd.h:
```diff
@@ -69,6 +70,9 @@ struct Config {
   bool adapter_monitoring_enabled = false;
 
+  bool wait_disk_online = false;
+  int wait_online_timeout = DEFAULT_IMAGE_MAP_TIMEOUT;
```
... it also seems like there shouldn't be a timeout on this step.  I see that there are some existing cumulative timeouts in service mode, but it doesn't seem like the plain `map <imagespec>` command has any.  Can the "wait online" step be covered by some existing timeout in service mode, leaving the plain command as is?

petrutlucian94 replied:
Right, adding yet another timeout option might seem redundant. We do need to have a timeout though, we can't poll indefinitely. What we could do would be to substract the time it took to issue the map request from the configured map timeout.

We're currently using map timeouts in a few places:

* when the map command issues the request to the centralized service through a named pipe: https://github.com/ceph/ceph/blob/4af9db9102780c7eef05d154f1148c682c8e1603/src/tools/rbd_wnbd/rbd_wnbd.cc#L300-L307
* when the centralized service spawns the map subprocess: https://github.com/ceph/ceph/blob/4af9db9102780c7eef05d154f1148c682c8e1603/src/tools/rbd_wnbd/rbd_wnbd.cc#L785
* when the service restarts: https://github.com/ceph/ceph/blob/4af9db9102780c7eef05d154f1148c682c8e1603/src/tools/rbd_wnbd/rbd_wnbd.cc#L584-L587

The problem is that in the first two situations we're using the default hard coded value instead of honoring the configurable value. There's a TODO, can't remember if there was a specific reason why we kept using the hardcoded value.

The "wait online" step can be performed either on the caller side (as proposed by this patch) and/or on the service side. I'm a bit concerned about performing the wait when the service starts. For example, the service start time may increase significantly if we wait for 100 disks.

I'd revisit the timeout handling after we merge this: https://github.com/ceph/ceph/pull/52540.

petrutlucian94 replied:
I've updated the PR, dropping the new parameters and performing the wait on the service side.

idryomov replied:
> On the other hand, it might be the right thing to do, guaranteeing that all rbd mappings are available once the service is running.

Yeah, it's why the service is there after all.

I'm not opposed to optionally skipping the wait, but I strongly believe that the default behavior should be to wait -- especially for a case of mapping a single disk (e.g. what fsx does).

idryomov commented on src/tools/rbd_wnbd/rbd_wnbd.cc:
```diff
@@ -1296,6 +1323,13 @@ static int do_map(Config *cfg)
   }
 
+  // TODO: consider substracting the time it took to perform the
+  // above operations from the following timeout.
```
It's not immediately clear what the "following timeout" is.  I would suggest rephrasing or making `wait_mapped_disk` take a timeout argument.
```suggestion
  // above operations from cfg->image_map_timeout in
  // wait_mapped_disk().
```

idryomov commented on src/tools/rbd_wnbd/rbd_wnbd.cc:
```diff
@@ -1210,7 +1232,12 @@ static int do_map(Config *cfg)
   if (g_conf()->daemonize && cfg->parent_pipe.empty()) {
-    return send_map_request(get_cli_args());
+    r = send_map_request(get_cli_args());
+    if (r) {
```
```suggestion
    if (r < 0) {
```

petrutlucian94 replied:
Out of caution, I'd check for any non-zero value here since we may end up forwarding a Windows status code.

idryomov commented on src/tools/rbd_wnbd/rbd_wnbd.cc:
```diff
@@ -1296,6 +1323,13 @@ static int do_map(Config *cfg)
+  // TODO: consider substracting the time it took to perform the
+  // above operations from the following timeout.
+  r = wait_mapped_disk(cfg);
+  if (r) {
```
```suggestion
  if (r < 0) {
```

petrutlucian94 commented:
> You have checked "Includes unit test(s)" in the PR description but I don't see any. Did you mean integration tests -- perhaps that this would be effectively covered by fsx?

There was a small change in the fsx test (the wait parameter), I'll uncheck it.

petrutlucian94 commented:
The Windows service fails to start for some reason (works locally). I've updated the command that starts the service to get additional information: https://github.com/ceph/ceph-build/pull/2162/.

petrutlucian94 commented:
jenkins test windows

idryomov commented:
> The Windows service fails to start for some reason (works locally). I've updated the command that starts the service to get additional information: [ceph/ceph-build#2162](https://github.com/ceph/ceph-build/pull/2162).

```
+ timeout 30s ssh -i **** -o UserKnownHostsFile=/home/jenkins-build/build/workspace/ceph-windows-pull-requests/libvirt/known_hosts administrator@192.168.122.154 powershell.exe net.exe start ceph-rbd
The service is not responding to the control function.

More help is available by typing NET HELPMSG 2186.

+ EXIT_CODE=1
+ [[ 1 -eq 124 ]]
+ return 1
```

petrutlucian94 commented on src/tools/rbd_wnbd/rbd_wnbd.cc:
```diff
@@ -482,6 +482,13 @@ int map_device_using_suprocess(std::string arguments, int timeout_ms)
+        exit_code = -ECHILD;
+      } else if (exit_code > 0) {
+        // Make sure to return a negative error code.
+        exit_code *= -1;
```
Right, ``exit_code = -exit_code`` is more readable, I'll go with that.

petrutlucian94 commented:
> > The Windows service fails to start for some reason (works locally). I've updated the command that starts the service to get additional information: [ceph/ceph-build#2162](https://github.com/ceph/ceph-build/pull/2162).
> 
> ```
> + timeout 30s ssh -i **** -o UserKnownHostsFile=/home/jenkins-build/build/workspace/ceph-windows-pull-requests/libvirt/known_hosts administrator@192.168.122.154 powershell.exe net.exe start ceph-rbd
> The service is not responding to the control function.
> 
> More help is available by typing NET HELPMSG 2186.
> 
> + EXIT_CODE=1
> + [[ 1 -eq 124 ]]
> + return 1
> ```

Yep, it's probably crashing because it's using an out of date WNBD version. I forgot that the Windows image build job is using the ceph wnbd fork: https://github.com/ceph/ceph-build/blob/71bf6746bb90692075e245ff08cc11da2a09e0e5/ceph-windows-image-build/build/setup.ps1#L12-L13

I've updated the fork and triggered a new image build: https://jenkins.ceph.com/view/Windows/job/ceph-windows-image-build/29/console

petrutlucian94 commented:
jenkins test windows

petrutlucian94 commented:
jenkins test make check

petrutlucian94 commented:
jenkins test api

petrutlucian94 commented on src/tools/rbd_wnbd/rbd_wnbd.cc:
```diff
@@ -482,6 +482,13 @@ int map_device_using_suprocess(std::string arguments, int timeout_ms)
   clean_process:
     if (!is_process_running(pi.dwProcessId)) {
       GetExitCodeProcess(pi.hProcess, (PDWORD)&exit_code);
+      if (!exit_code) {
```
Right, those assignments are indeed redundant.

``GetExitCodeProcess`` is expected to return the exit code if the process terminated (which we check using ``is_process_running``). We could also check the ``GetExitCodeProcess`` return value, zero means that the function failed and the actual error must be retrieved using ``GetLastError``, but we're already setting ``-ECHILD`` if the ``exit_code`` value is zero.

For what is worth, this function is expected to be removed once we switch to a single rbd-wnbd process per host: https://github.com/ceph/ceph/pull/52540.

idryomov replied:
> For what is worth, this function is expected to be removed once we switch to a single rbd-wnbd process per host: #52540.

I'll let it go then ;)
